src/itensor.jl

```diff
@@ -221,7 +213,7 @@ itensor(T::Tensor) = itensor(store(T), inds(T))
 Convert the ITensor to a Tensor that shares the same
 storage and indices as the ITensor.
 """
-tensor(A::ITensor) = tensor(store(A),inds(A))
+tensor(A::ITensor) = tensor(store(A), inds(A))
```

Comment:
So when we change the storage of `IndexSet` to `Vector`, this conversion becomes subtle (it's something that kept bothering me when I was thinking about this design). There is a tension between `NDTensors` and `ITensors`, where the `IndexSet` acts more like a `Tuple` within `NDTensors` (most of the operations are supposed to be type stable with respect to the tensor order in `NDTensors`) while in `ITensors` the `IndexSet` acts more like an ordered set.

In this new design, if we don't do any conversion of the `IndexSet` here, we are basically passing a `Vector` (i.e. a non-statically sized container) to `NDTensors`. I'm not sure what implications this has for performance within `NDTensors`, since I designed `NDTensors` to be generic with respect to the storage of the indices but had in mind that it would be statically sized, so likely I use that assumption in many places and using a non-statically sized container either doesn't work or leads to type unstable code.

The "simple" solution is to convert the indices to a `Tuple` here, i.e. ` tensor(A::ITensor) = tensor(store(A), Tuple(inds(A)))`, and then the conversion back to an `ITensor` from a `Tensor` would have to convert back to a `Vector`/`IndexSet`. However, this can have a decent amount of overhead if done naively (at least in the `Vector -> Tuple` direction, the `Tuple -> Vector` direction is fast):
```julia
julia> @btime Tuple($([prime(i, n) for n in 0:4]));
  256.702 ns (7 allocations: 1.69 KiB)

julia> @btime Tuple($([prime(i, n) for n in 0:12]));
  569.027 ns (15 allocations: 4.20 KiB)

julia> @btime Tuple($([prime(i, n) for n in 0:16]));
  734.744 ns (19 allocations: 5.61 KiB)

julia> @btime collect($(Tuple([prime(i, n) for n in 0:16])));
  117.304 ns (1 allocation: 1.98 KiB)
```

Interestingly, I can make a faster `Tuple` constructor like this:
```julia
@eval struct Val2{x}
  (f::Type{<:Val2})() = $(Expr(:new, :f))
end

_NTuple(::Val2{N}, v::Vector{T}) where {N, T} = ntuple(n -> v[n], Val(N))
_Tuple(v::Vector{T}) where {T} = _NTuple(Val2{length(v)}(), v)
```
using a faster `Val` as discussed [here](https://github.com/JuliaLang/julia/issues/36384) and I get the following:
```julia
julia> VERSION
v"1.5.3"

julia> @btime _Tuple($([prime(i, n) for n in 0:16]));
  243.172 ns (1 allocation: 1.77 KiB)
```
so we can make this conversion pretty fast. Ultimately I think converting to a `Tuple` before passing it to `NDTensors` would be the right way to go here.

Also note that the performance without the `Val2` hack is better in Julia v1.6:
```julia
julia> VERSION
v"1.6.0-rc2"

julia> @btime Tuple($([prime(i, n) for n in 0:16]));
  585.672 ns (18 allocations: 3.62 KiB)

julia> _NTuple(::Val{N}, v::Vector{T}) where {N, T} = ntuple(n -> v[n], Val(N));

julia> _Tuple(v::Vector{T}) where {T} = _NTuple(Val(length(v)), v);

julia> @btime _Tuple($([prime(i, n) for n in 0:16]));
  308.824 ns (1 allocation: 1.77 KiB)
```
but still better with the `Val2` trick:
```julia
julia> _NTuple(::Val2{N}, v::Vector{T}) where {N, T} = ntuple(n -> v[n], Val(N));

julia> _Tuple(v::Vector{T}) where {T} = _NTuple(Val2{length(v)}(), v);

julia> @btime _Tuple($([prime(i, n) for n in 0:16]));
  222.427 ns (1 allocation: 1.77 KiB)
```

So TLDR: I think we should convert the `IndexSet` to a `Tuple` here, using the specialized `_Tuple` function I wrote above that makes use of `Val2` (or better yet `Order`, which is defined like `Val2` and so is fast when being constructed from runtime values).

Reply:
It can be easy to lose context about what these timings mean, so for some reference here are benchmarks for small matrix multiplications performed with Julia Arrays and ITensors:
```julia
julia> n = 10; A = randn(n, n); @btime $A * $A;
  281.037 ns (1 allocation: 896 bytes)

julia> n = 20; A = randn(n, n); @btime $A * $A;
  733.406 ns (1 allocation: 3.25 KiB)

julia> n = 30; A = randn(n, n); @btime $A * $A;
  1.794 μs (1 allocation: 7.19 KiB)

julia> n = 40; A = randn(n, n); @btime $A * $A;
  2.958 μs (1 allocation: 12.63 KiB)

julia> n = 10; i = Index(n); A = randomITensor(i', i); @btime $(A') * $A;
  830.237 ns (22 allocations: 2.92 KiB)

julia> n = 20; i = Index(n); A = randomITensor(i', i); @btime $(A') * $A;
  1.364 μs (22 allocations: 5.30 KiB)

julia> n = 30; i = Index(n); A = randomITensor(i', i); @btime $(A') * $A;
  2.480 μs (22 allocations: 9.23 KiB)

julia> n = 40; i = Index(n); A = randomITensor(i', i); @btime $(A') * $A;
  3.235 μs (22 allocations: 14.67 KiB)
```
My goal is always to keep the gap between these down and making sure the conversions between ITensors and Tensors is fast is an important part of that (in fact you can see it can easily dominate the cost of small tensor contractions).

Reply:
Thanks, this is very helpful.

Reply:
Another reference benchmark:
```julia
julia> julia> VERSION
v"1.5.3"

julia> using ITensors.NDTensors

julia> i = Index(2, "i");

julia> A = randomITensor(i', i);

julia> @btime tensor($A);
  21.026 ns (1 allocation: 224 bytes)

julia> @btime itensor($(tensor(A)));
  21.543 ns (3 allocations: 272 bytes)
```
so conversion between ITensors and Tensors is very fast right now. Unfortunately with this new design (converting `IndexSet -> Tuple`) this conversion will slow down a bit, but I think the bigger picture is that it will make a lot of `IndexSet` operations faster and more type stable and our focus is more on larger tensor contractions.

Reply:
Agreed. Anecdotally in code I've worked on I've been slowed down *a lot* by the first run of `contract` or `norm` or even `ndims` because of type inference issues.

Reply:
I'm surprised by `ndims`, but yes any code that involves contractions and decompositions has very slow compilation right now. I started analyzing it with SnoopCompile and found that the NDTensors `contract`, `svd` and related functions take up a ton of inference time. I think the key things are to make sure the code in NDTensors is type stable and make better precompile statements for those functions with many combinations of tensor orders and storage types.

Reply:
💯 for SnoopCompile, great package.

Reply:
Going down a bit of a rabbit hole, but here is an even faster Tuple constructor based on the caching strategy explained [here](https://github.com/JuliaLang/julia/issues/29887):
```julia
const ValCacheLength = 100
const ValCache = [Val(n) for n in 1:ValCacheLength]
_NTuple(::Val{N}, v::Vector{T}) where {N, T} = ntuple(n -> v[n], Val(N))
_Tuple(v::Vector{T}) where {T} = _NTuple(ValCache[length(v)], v)
```
results in:
```julia
julia> @btime _Tuple($([prime(i, n) for n in 0:16]));
  165.160 ns (1 allocation: 1.77 KiB)
```
so this is the fastest version yet. The main downside is that it puts an effective limit on the order of the tensors, but I don't see a downside in making the limit quite large (and an order-100 tensor is already well beyond any we would expect to see).

Reply:
As long as the limit can be adjusted later (in the unlikely event we do hit such tensors) it seems fine to me.